File: src/netlify/functions/badge.js
export async function handler(event, context) {
	const path = event.path; // e.g. "/function/hello"
	const params = new URLSearchParams(event.rawQuery || "");

	let name = path.split("/")[3];
	const noLogo = params.get("noLogo") || false;


	const badgeLookup = await (
		await fetch(`${process.env.URL}/badgeList.json`)
	).json();

	let badgeData;
	for (const cate of Object.keys(badgeLookup)) {

		badgeData = badgeLookup[cate][name] || null
		if (badgeData != null) {
			break
		}
	}


	if (badgeData == null) {
		return {
			statusCode: 404,
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify({
				error: `Unable to find badge with id ${name}`,
			}),
		};
	}

	const response = await fetch(
		`${process.env.URL}/api/customBadge?name=${badgeData.name}&textColor=${badgeData.textColor}&bgColor=${badgeData.bgColor}&image=${process.env.URL}${badgeData.image}&noLogo=${noLogo}`
	);

	const data = await response.text();

	return {
		statusCode: 200,
		headers: { "Content-Type": "image/svg+xml", "Cache-Control": "public, max-age=3600" },
		body: data,
	};
}

File: src/netlify/functions/badge.py
import json
from pydantic_extra_types import Color
from netlify.functions.custom_badge import custom_badge

badge_list = {}
with open("badgeList.json", "r") as f:
    badge_list = json.load(f)

def deep_get(dictionary, target_key):
    if target_key in dictionary:
        return dictionary[target_key]
    
    for _, value in dictionary.items():
        for key2, value2 in value.items():
            if key2 == target_key:
                return value2

    return {}

async def badge(
        name: str,
        no_logo: bool = False,
        rounded: bool = False
    ):

    badge_info = deep_get(badge_list, name)

    return await custom_badge(
        name=badge_info.get("name"),
        textColor=Color(badge_info.get("textColor", "#FFFFFF")),
        backgroundColor=Color(badge_info.get("bgColor", "#000000")),
        image=f'/static{badge_info.get("image")}',
        no_logo=no_logo,
        rounded=rounded
    )
    